src/scenic/core/visibility.py

```diff
@@ -0,0 +1,503 @@
+    """
+    from scenic.core.object_types import Point, OrientedPoint, Object
+
+    # Filter occluding objects that are obviously infeasible
+    occludingObjects = [obj for obj in occludingObjects if position.distanceTo(obj) <= visibleDistance]
+
+    if rayCount is None:
+        rayCount = (math.degrees(viewAngles[0])*rayDensity, math.degrees(viewAngles[1])*rayDensity)
+
+        if distanceScaling:
+            target_distance = target.position.distanceTo(position)
+
+            rayCount = (rayCount[0]*viewAngles[0]*target_distance, rayCount[1]*viewAngles[1]*target_distance)
+
+        altitudeScaling = True
+    else:
+        # Do not scale ray counts with altitude or distance if explicitly given
+        altitudeScaling = False
+
+    if isinstance(target, (Region, Object)):
+        # Extract the target region from the object or region.
+        if isinstance(target, Region):
+            raise NotImplementedError
+        elif isinstance(target, Object):
+            # If the object contains its center and we can see the center, the object
+            # is visible.
+            if target.shape.containsCenter and \
+                canSee(position, orientation, visibleDistance, viewAngles,
+                       rayCount, rayDensity, distanceScaling,
+                       target.position, occludingObjects):
+                return True
+            target_region = target.occupiedSpace
+
+        # Check that the distance to the target is not greater than visibleDistance, and
+        # see if position is in the target.
+        if target.containsPoint(position):
+            return True
+
+        if target.distanceTo(position) > visibleDistance:
+            return False
+
+        # Orient the object so that it has the same relative position and orientation to the
+        # origin as it did to the viewer
+        target_vertices = target_region.mesh.vertices - np.array(position.coordinates)
+
+        if orientation is not None:
+            target_vertices = orientation.inverse.getRotation().apply(target_vertices)
+
+        # Add additional points along each edge that could potentially have a higher altitude
+        # than the endpoints.
+        vec_1s = np.asarray(target_vertices[target_region.mesh.edges[:,0],:])
+        vec_2s = np.asarray(target_vertices[target_region.mesh.edges[:,1],:])
+        x1, y1, z1 = vec_1s[:,0], vec_1s[:,1], vec_1s[:,2]
+        x2, y2, z2 = vec_2s[:,0], vec_2s[:,1], vec_2s[:,2]
+        D = x1*x2 + y1*y2
+        N = (x1**2 + y1**2)*z2 - D*z1
+        M = (x2**2 + y2**2)*z1 - D*z2
+        with np.errstate(divide='ignore', invalid='ignore'):
+            t_vals = N/(N+M) # t values that can be an altitude local optimum
+
+        # Keep only points where the t_value is between 0 and 1
+        t_mask = np.logical_and(t_vals > 0, t_vals < 1)
+        interpolated_points = vec_1s[t_mask] + t_vals[t_mask][:,None] * (vec_2s[t_mask] - vec_1s[t_mask])
```

Comment:
You use the difference `vec_2s - vec_1s` below, so might as well compute that only once and mask it here.